src/CrossmintEmbed.ts

```diff
@@ -0,0 +1,225 @@
+import buildSiteMetadata from "./utils/siteMetadata";
+import { sleep } from "./utils/sleep";
+
+export default class CrossmintEmbed {
+    private _config: CrossmintEmbedConfig;
+
+    private get _frameUrl() {
+        return this._config.environment + "/frame";
+    }
+
+    private constructor(config: CrossmintEmbedConfig) {
+        console.log("[crossmint-embed] Initialized embed with version:", config.libVersion);
+
+        this._config = config;
+    }
+
+    static async init(config: CrossmintEmbedConfig) {
+        const client = new CrossmintEmbed(config);
+
+        return client;
+    }
+
+    async login(): Promise<string | undefined | null> {
+        const account = await this.getLoginFromIFrame();
+
+        if (account !== undefined) {
+            console.log("[crossmint] Received account from auto connect");
+            return account;
+        }
+
+        const crossmintWindow = new WindowAdapter();
+        await crossmintWindow.init({ parentWindow: window, url: this._frameUrl });
+
+        return await new Promise<string | undefined | null>(async (resolve, reject) => {
+            console.log("[crossmint-embed] Waiting login");
+
+            let _account: string | undefined | null = undefined;
+
+            const handleMessage = async (e: MessageEvent<any>) => {
+                if (!ALLOWED_ORIGINS.includes(e.origin)) return;
+
+                const { request, data } = e.data;
+
+                switch (request) {
+                    case CrossmintEmbedRequestType.REQUEST_ACCOUNTS:
+                        const { account } = data as { account: string };
+
+                        _account = account;
+                        crossmintWindow.controlledWindow?.close();
+                        break;
+                    case CrossmintEmbedRequestType.USER_REJECT:
+                        console.log("[crossmint-embed] User rejected login");
+                        _account = null;
+                        break;
+                    default:
+                        break;
+                }
+            };
+
+            window.addEventListener("message", handleMessage);
+
+            while (crossmintWindow.open && crossmintWindow.controlledWindow) {
+                this.postMessage(
+                    crossmintWindow.controlledWindow,
+                    CrossmintEmbedRequestType.REQUEST_ACCOUNTS,
+                    undefined,
+                    this._frameUrl
+                );
+
+                await sleep(100);
+            }
+
+            window.removeEventListener("message", handleMessage);
+            resolve(_account);
+        });
+    }
+
+    async signMessage(message: Uint8Array): Promise<Uint8Array | undefined | null> {
+        const crossmintWindow = new WindowAdapter();
+        await crossmintWindow.init({ parentWindow: window, url: this._frameUrl });
+
+        return await new Promise<Uint8Array | undefined | null>(async (resolve, reject) => {
```

Comment:
90% of the logic here is the same as the Promise on line 43 above, can you extract it somewhere?

Reply:
Filed a task https://linear.app/crossmint/issue/ENG-799/[sdk]-extract-logic-for-awaiting-postmessage-interaction